### ISSUE TITLE
FluxC integration: Better error management in SignInFragment

### ISSUE THREAD

**Issue by maxme** [wordpress-mobile/WordPress-Android#4641]
While reviewing https://github.com/wordpress-mobile/WordPress-Android/pull/4430, I noticed we could get `stream canceled` volley request when using a very slow network (throttled), these errors are ignored during signin in SignInFragment, and the user is stuck on that screen with a progress bar and can't do nothing.

We should make sure 1. FluxC generate onChanged events with the error field filled and 2. all errors are treated in SignInFragment.

**Comment by oguzkocer** [comment]
I've been testing the login flow in `feature/fluxc-integration` and most paths seems to be working great. I did come across a few edge cases, which I will document in separate comments. I've tried logging into a .com account, self-hosted account, magic login and many error paths which were all working as expected.

Here are the error paths that are working as expected:

When I used a wrong password:

![wrong password](https://cloud.githubusercontent.com/assets/662023/23296852/fdafaec0-fa87-11e6-9713-c21891e6c204.png)

When I entered wrong url for self-hosted site:

![wrong url](https://cloud.githubusercontent.com/assets/662023/23296871/0dfd4ea4-fa88-11e6-9015-2ebc7297a96e.png)

When I exceeded the login limit:

![login limit](https://cloud.githubusercontent.com/assets/662023/23296896/256721c8-fa88-11e6-8a78-440d4f7efd5b.png)

When I turned off the network in the middle of the call:

![turn off network](https://cloud.githubusercontent.com/assets/662023/23296908/34e3c656-fa88-11e6-94a5-16f20c06cbe5.png)

**Comment by oguzkocer** [comment]
I wanted to test throttling and slow network connections, so I've changed my emulator to use the `gsm` speed as documented here: https://developer.android.com/studio/run/emulator-commandline.html#netspeed Everything worked just fine with this speed. Things were a bit slow to show up, which is expected, but no request failed due to timeout or nothing unexpected happened really. So, I thought I'd turn off the network during the login call, which work expected once and broke in my other attempt.

I think this is a really big edge case, but here is what I think happened. Just as I turned off the network, I think the request finished but the UI didn't get updated, because I was on settings, so the app got stuck like this:

![logged in but not](https://cloud.githubusercontent.com/assets/662023/23297141/62f9dce6-fa89-11e6-8c09-f5c0f2515433.png)

So basically I got stuck in the login screen. Restarting the app worked fine as I was logged in as expected. I am not really sure if it's worth spending our time trying to fix this edge case. As I mentioned in my previous comment, if the request is not yet finished, we get a "We can't log you in" error.

**Comment by oguzkocer** [comment]
_The following didn't turn out to be an issue, but keeping it for documentation purposes:_

One error case I was able to reproduce consistently is the following:

Username: {anything}
Password: {anything}
Url: oguzkocer.sandbox.me

![stuck](https://cloud.githubusercontent.com/assets/662023/23297422/8113c858-fa8a-11e6-824e-51c4eec9a105.png)

This is a case I came across by mistake. I have a self-hosted site with the following url: `oguzkocer.wpsandbox.me`. I was trying to try the wrong password for a self-hosted site and I mistyped my url and forgot the `wp` part and the login just got stuck. Here are the relevant logs from the attempt:

```
I/WordPress-NUX: User tries to sign in on Self Hosted: oguzkocer.sandbox.me with username: someusername
D/WordPress-API: Dispatching action: AuthenticationAction-DISCOVER_ENDPOINT
I/WordPress-NUX: Calling system.listMethods on the following URLs: [https://oguzkocer.sandbox.me/xmlrpc.php, http://oguzkocer.sandbox.me/xmlrpc.php, https://oguzkocer.sandbox.me, http://oguzkocer.sandbox.me, oguzkocer.sandbox.me]
I/WordPress-NUX: Trying system.listMethods on the following URL: https://oguzkocer.sandbox.me/xmlrpc.php
E/WordPress-API: Couldn't get XML-RPC response.
E/WordPress-NUX: The response of system.listMethods was empty for https://oguzkocer.sandbox.me/xmlrpc.php
I/WordPress-NUX: Trying system.listMethods on the following URL: http://oguzkocer.sandbox.me/xmlrpc.php
E/Volley: [170] BasicNetwork.performRequest: Unexpected response code 404 for http://oguzkocer.sandbox.me/xmlrpc.php
E/WordPress-API: Volley error
                 com.android.volley.ServerError
                     at com.android.volley.toolbox.BasicNetwork.performRequest(BasicNetwork.java:179)
                     at com.android.volley.NetworkDispatcher.run(NetworkDispatcher.java:114)
E/WordPress-NUX: The response of system.listMethods was empty for http://oguzkocer.sandbox.me/xmlrpc.php
I/WordPress-NUX: Trying system.listMethods on the following URL: https://oguzkocer.sandbox.me
E/WordPress-API: Volley error
                 com.android.volley.TimeoutError
                     at com.android.volley.toolbox.BasicNetwork.performRequest(BasicNetwork.java:147)
                     at com.android.volley.NetworkDispatcher.run(NetworkDispatcher.java:114)
E/WordPress-API: Couldn't get XML-RPC response.
E/WordPress-NUX: The response of system.listMethods was empty for https://oguzkocer.sandbox.me
I/WordPress-NUX: Trying system.listMethods on the following URL: http://oguzkocer.sandbox.me
E/WordPress-API: Can't deserialize XMLRPC response
                 org.xmlpull.v1.XmlPullParserException: Unexpected token (position:TEXT a-href="page:U74...@1:40 in java.io.InputStreamReader@b409dba) 
                     at org.kxml2.io.KXmlParser.next(KXmlParser.java:432)
                     at org.kxml2.io.KXmlParser.next(KXmlParser.java:313)
                     at org.kxml2.io.KXmlParser.nextTag(KXmlParser.java:2050)
                     at org.wordpress.android.fluxc.network.xmlrpc.XMLSerializerUtils.deserialize(XMLSerializerUtils.java:63)
                     at org.wordpress.android.fluxc.network.xmlrpc.XMLRPCRequest.parseNetworkResponse(XMLRPCRequest.java:60)
                     at com.android.volley.NetworkDispatcher.run(NetworkDispatcher.java:125)
E/WordPress-API: Volley error
                 com.android.volley.ParseError: org.xmlpull.v1.XmlPullParserException: Unexpected token (position:TEXT a-href="page:U74...@1:40 in java.io.InputStreamReader@b409dba) 
                     at org.wordpress.android.fluxc.network.xmlrpc.XMLRPCRequest.parseNetworkResponse(XMLRPCRequest.java:71)
                     at com.android.volley.NetworkDispatcher.run(NetworkDispatcher.java:125)
                  Caused by: org.xmlpull.v1.XmlPullParserException: Unexpected token (position:TEXT a-href="page:U74...@1:40 in java.io.InputStreamReader@b409dba) 
                     at org.kxml2.io.KXmlParser.next(KXmlParser.java:432)
                     at org.kxml2.io.KXmlParser.next(KXmlParser.java:313)
                     at org.kxml2.io.KXmlParser.nextTag(KXmlParser.java:2050)
                     at org.wordpress.android.fluxc.network.xmlrpc.XMLSerializerUtils.deserialize(XMLSerializerUtils.java:63)
                     at org.wordpress.android.fluxc.network.xmlrpc.XMLRPCRequest.parseNetworkResponse(XMLRPCRequest.java:60)
                     at com.android.volley.NetworkDispatcher.run(NetworkDispatcher.java:125) 
E/WordPress-NUX: The response of system.listMethods was empty for http://oguzkocer.sandbox.me
E/WordPress-NUX: Invalid URL: oguzkocer.sandbox.me
E/WordPress-NUX: system.listMethods failed for oguzkocer.sandbox.me
                 org.wordpress.android.fluxc.network.discovery.SelfHostedEndpointFinder$DiscoveryException
                     at org.wordpress.android.fluxc.network.discovery.SelfHostedEndpointFinder.doSystemListMethodsXMLRPC(SelfHostedEndpointFinder.java:405)
                     at org.wordpress.android.fluxc.network.discovery.SelfHostedEndpointFinder.checkXMLRPCEndpointValidity(SelfHostedEndpointFinder.java:364)
                     at org.wordpress.android.fluxc.network.discovery.SelfHostedEndpointFinder.verifyXMLRPCUrl(SelfHostedEndpointFinder.java:187)
                     at org.wordpress.android.fluxc.network.discovery.SelfHostedEndpointFinder.verifyOrDiscoverXMLRPCEndpoint(SelfHostedEndpointFinder.java:133)
                     at org.wordpress.android.fluxc.network.discovery.SelfHostedEndpointFinder.access$100(SelfHostedEndpointFinder.java:40)
                     at org.wordpress.android.fluxc.network.discovery.SelfHostedEndpointFinder$1.run(SelfHostedEndpointFinder.java:108)
                     at java.lang.Thread.run(Thread.java:761)
W/WordPress-NUX: The XML-RPC endpoint was not found by using our 'smart' cleaning approach. Time to start the Endpoint discovery process
I/WordPress-NUX: Running RSD discovery process on the following URLs: [oguzkocer.sandbox.me, https://oguzkocer.sandbox.me, http://oguzkocer.sandbox.me]
I/WordPress-NUX: Downloading the HTML content at the following URL: https://oguzkocer.sandbox.me
E/WordPress-API: Volley error
                 com.android.volley.TimeoutError
                     at com.android.volley.toolbox.BasicNetwork.performRequest(BasicNetwork.java:147)
                     at com.android.volley.NetworkDispatcher.run(NetworkDispatcher.java:114)
E/WordPress-API: Couldn't get XML-RPC response
W/WordPress-NUX: Content downloaded but it's empty or null. Skipping this URL
I/WordPress-NUX: Downloading the HTML content at the following URL: http://oguzkocer.sandbox.me
I/WordPress-NUX: Can't find the RSD endpoint in the HTML document. Try to check the pingback tag, and the apiLink tag.
I/WordPress-NUX: XML-RPC endpoint not found
D/WordPress-API: Dispatching action: AuthenticationAction-DISCOVERY_RESULT
E/WordPress-API: onDiscoveryResponse has error: NO_SITE_ERROR - NO_SITE_ERROR
E/WordPress-API: Discover error: NO_SITE_ERROR
```

As I typed all of this, it did actually threw an error. So, I guess things are working, but waiting for that long felt like it's stuck. Here is the error I got in the end:

![login error timeout](https://cloud.githubusercontent.com/assets/662023/23297778/df5c9cc2-fa8b-11e6-8d83-9bc7221d369e.png)

**Comment by oguzkocer** [comment]
The above issue was happening on the emulator consistently, but it's working as expected on my device and @maxme couldn't reproduce either. So, no need to pursue it.

**Comment by oguzkocer** [comment]
Closing the issue, since we didn't find any issues during tests.
